src/DBFTables.jl

```diff
@@ -1,150 +1,293 @@
+
+Read a source, a path to a file or an opened stream, to a DBFTables.Table.
+This type conforms to the Tables interface, so it can be easily converted
+to other formats. It is possible to iterate through the rows of this object,
+or to retrieve columns like `dbf.fieldname`.
+"""
+function Table(io::IO)
+    header = Header(io)
+    # consider using mmap here for big dbf files
+    data = Vector{UInt8}(undef, header.rsize * header.records)
+    read!(io, data)
+    strings = _create_stringarray(header, data)
+    Table(header, data, strings)
+end
+
+function Table(path::AbstractString)
+    open(path) do io
+        Table(io)
+    end
+end
+
+"Collect all the offsets and lenghts from the header to create a StringArray"
+function _create_stringarray(header::Header, data::AbstractVector)
+    # first make the lengths and offsets for a single record
+    lengths_record = UInt32.(getfield.(header.fields, :length))
+    offsets_record = vcat(0, cumsum(lengths_record)[1:end-1]) .+ 1
+
+    # the lengths are equal for each record
+    lengths = repeat(lengths_record, 1, header.records)
+    # the offsets accumulate over records with the record size
+    row_offsets = range(0; length = header.records, step = header.rsize)
+    offsets = repeat(offsets_record, 1, header.records)
+    offsets .+= reshape(row_offsets, 1, :)
+
+    StringArray{WeakRefString{UInt8},2}(data, offsets, lengths)
+end
+
+"Create a NamedTuple representing a single row"
+function Base.NamedTuple(row::Row)
+    dbf = gettable(row)
+    str = getstrings(dbf)
+    fields = getfields(dbf)
+    ncol = length(fields)
+    rowidx = getrow(row)
+    @inbounds record = @view str[:, rowidx]
+    @inbounds prs = (fields[col].name => dbf_value(
+        fields[col].type,
+        record[col],
+    ) for col = 1:ncol)
+    return (; prs...)
+end
+
+function Base.show(io::IO, row::Row)
+    show(io, NamedTuple(row))
+end
+
+function Base.show(io::IO, dbf::Table)
+    nr = length(dbf)
+    nc = length(getfields(dbf))
+    println(io, "DBFTables.Table with $nr rows and $nc columns")
+end
+
+Base.isempty(dbf::Table) = getheader(dbf).records == 0
+
+"Get a BitVector which is true for rows that are marked as deleted"
+function isdeleted(dbf::Table)
+    data = getfield(dbf, :data)
+    rsize = getheader(dbf).rsize
+    nrow = getheader(dbf).records
+    idx = range(1, step = rsize, length = nrow)
+    data[idx] .== 0x2a
+end
+
+"Check if the row is marked as deleted"
+function isdeleted(dbf::Table, row::Integer)
+    data = getfield(dbf, :data)
+    i = (row - 1) * getheader(dbf).rsize + 1
+    data[i] == 0x2a
+end
+
+"Iterate over the rows of a DBF Table, yielding a DBFTables.Row for each row"
+function Base.iterate(dbf::Table, st = 1)
+    st > length(dbf) && return nothing
+    return Row(dbf, st), st + 1
+end
+
+function Base.getproperty(row::Row, name::Symbol)
+    dbf = gettable(row)
+    str = getstrings(dbf)
+    colidx = findfirst(x -> x === name, propertynames(dbf))
```

Comment:
In CSV.jl, I recently added a `lookup::Dict{Symbol, Int}` to the `CSV.File` object due to complains that doing `findfirst` in `getproperty` was too slow. It's pretty cheap/simple to throw the `lookup` into the `Table` or `Header` type and use that in the `getproperty` methods.

Reply:
Ah yes, that's a good tip, especially for if there are many columns. Done in 01fbb419936a2b6dd85fa039f47e8187bd4ee6f5